File: back/runtime/endpointfitness/endpointfitness.go
// Package endpointfitness is the DP08 arch-fitness invariant
// EMITTED_NO_HARDCODED_ENDPOINT made an ENFORCED, deterministic build-time
// sensor over the EMITTED TS tree (Hono handlers, workers, boot config —
// ROADMAP-provisioning-deploy EPIC B): no host, URL, IP or concrete
// host:port may live as a literal in emitted source; every endpoint flows
// through the DP07 projection (resolveConnection → process.env / requireEnv
// at boot — SPEC-stack-2026: « AUCUNE URL/secret en dur — tout par variables
// d'environnement »).
//
// THE PASS (A7, determinism-first): detection is a LITERAL PARSE of the TS
// source — a deterministic lexer that walks strings, template literals and
// their ${expr} spans exactly the way the TS AST does (the authoritative TS
// twin front/web/lib/endpoint-fitness.ts runs the REAL TypeScript compiler
// API over the same rules; this Go pass is its byte-parity-pinned twin) —
// NEVER an LLM that "judges if it looks hardcoded" (ADR 0066). The four
// closed reasons:
//
//   - localhost_literal  — `localhost` / 127.0.0.1 / 0.0.0.0 in a literal;
//   - ip_literal         — a concrete IPv4 dotted-quad;
//   - url_concrete_host  — scheme://host where the authority's host carries
//     concrete characters (a fully env-derived host — ${VAR} or a template
//     ${expr} span — is allowed: the env poisons adjacency);
//   - host_port_literal  — a concrete host:port (the container-name
//     convention ${APP_NAME}-<svc>:<port> is env-derived, hence allowed).
//
// FAIL-CLOSED + THE RATCHET: a red verdict yields the actionable BlockReason
// EMITTED_NO_HARDCODED_ENDPOINT and reddens the S84 archfit sensor
// (SensorArchFit), so selfcert.Certify BLOCKS THE CUT before the build loop
// can claim green (anti-passthrough — a missing verdict is red). The declared
// rule lives above the line in back/runtime/agentloop/arch-fitness.json
// (widening or softening it is idée → miroir → /goal, never a silent edit);
// the parity fixture pins config ↔ code.
//
// PURE (CLAUDE.md §6/§8): no DB, no clock, no rng, no I/O — Sense is a total
// function of the emitted tree; the verdict is content-addressed via
// records.Hash(records.Canonicalize(...)) (S02 reused) so « même arbre →
// même verdict » is a pinned reproducibility mirror. THE WALL (§2): this
// package writes nothing — it reports; the fitness config is above the line.
package endpointfitness

import (
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"github.com/steph-frtech/aidos/back/kernel/records"
	"github.com/steph-frtech/aidos/back/runtime/blockreason"
	"github.com/steph-frtech/aidos/back/runtime/buildloop/selfcert"
)

// CodeEmittedNoHardcodedEndpoint is the DP08 invariant code — the
// arch-fitness rule name AND the BlockReason code of a refused cut
// (package-local, the FN04 motif: never a new member of the frozen
// blockreason registry).
const CodeEmittedNoHardcodedEndpoint blockreason.Code = "EMITTED_NO_HARDCODED_ENDPOINT"

// Reason is the CLOSED classification of one hardcoded-endpoint finding.
type Reason string

const (
	// ReasonLocalhost — `localhost`, 127.0.0.1 or 0.0.0.0 in an emitted literal.
	ReasonLocalhost Reason = "localhost_literal"
	// ReasonIPLiteral — a concrete IPv4 dotted-quad in an emitted literal.
	ReasonIPLiteral Reason = "ip_literal"
	// ReasonURLConcreteHost — scheme://host with a concrete (non-env) host.
	ReasonURLConcreteHost Reason = "url_concrete_host"
	// ReasonHostPort — a concrete host:port outside the env-derived convention.
	ReasonHostPort Reason = "host_port_literal"
)

var reasonOrder = []Reason{ReasonLocalhost, ReasonIPLiteral, ReasonURLConcreteHost, ReasonHostPort}

// Reasons returns the closed reason set in canonical order.
func Reasons() []Reason {
	out := make([]Reason, len(reasonOrder))
	copy(out, reasonOrder)
	return out
}

// Verdict states — the closed two-value sensor verdict (no third state).
const (
	// StateGreen — no hardcoded endpoint in the emitted tree.
	StateGreen = "green"
	// StateRed — at least one finding; the cut is blocked.
	StateRed = "red"
)

// Finding is ONE hardcoded-endpoint occurrence in the emitted tree.
type Finding struct {
	// File is the emitted file path carrying the literal.
	File string `json:"file"`
	// Line is the 1-based line of the literal's start.
	Line int `json:"line"`
	// Literal is the offending literal text (raw, as written).
	Literal string `json:"literal"`
	// Reason is the closed classification.
	Reason Reason `json:"reason"`
}

// Verdict is the DP08 sensor verdict over one emitted tree — pure data,
// content-addressable (the reproducibility mirror pins Address).
type Verdict struct {
	// Rule is always EMITTED_NO_HARDCODED_ENDPOINT.
	Rule string `json:"rule"`
	// State is green or red (closed).
	State string `json:"state"`
	// TreeAddress is the content address of the scanned tree (S02 canonical).
	TreeAddress string `json:"tree_address"`
	// Findings are the hardcoded occurrences, sorted (file, line, literal).
	Findings []Finding `json:"findings"`
}

// ---------------------------------------------------------------------------
// The literal pass — a deterministic TS lexer with AST-equivalent semantics.
// ---------------------------------------------------------------------------

// sentinel poisons adjacency where a template ${expr} span or a textual
// ${VAR} placeholder stood: an env-derived fragment can never complete a
// concrete host. NUL never appears in real TS source.
const sentinel = "\x00"

// Literal is one extracted literal: its AST-equivalent text (template parts
// joined with the sentinel) and its 1-based start line.
type Literal struct {
	Text string
	Line int
}

// ExtractLiterals lexes TS source and returns every string literal, every
// template literal (its raw text parts joined with the sentinel — the
// TemplateExpression semantics) and, exactly like the AST walk, the literals
// found INSIDE template ${expr} spans. Comments are skipped. Total: any
// input yields a deterministic result (EOF closes an open token).
func ExtractLiterals(src string) []Literal {
	var out []Literal
	line := 1
	i := 0
	n := len(src)

	// template tracks the enclosing template literals: each entry accumulates
	// raw parts; braces counts the open braces inside the current ${ } span.
	type tmpl struct {
		parts     []string
		startLine int
		braces    []int // stack of open-brace counts, one per nested ${ span
	}
	var stack []*tmpl

	flushTemplate := func(t *tmpl) {
		out = append(out, Literal{Text: strings.Join(t.parts, sentinel), Line: t.startLine})
	}

	readQuoted := func(quote byte) (string, int) {
		// i points AT the opening quote; returns raw content + start line.
		start := i + 1
		startLine := line
		j := start
		for j < n {
			c := src[j]
			if c == '\\' && j+1 < n {
				if src[j+1] == '\n' {
					line++
				}
				j += 2
				continue
			}
			if c == quote || c == '\n' {
				break
			}
			j++
		}
		raw := src[start:j]
		if j < n && src[j] == quote {
			j++
		}
		i = j
		return raw, startLine
	}

	for i < n {
		c := src[i]
		// inside a template's raw part?
		inTemplate := len(stack) > 0 && len(stack[len(stack)-1].braces) == 0

		if inTemplate {
			t := stack[len(stack)-1]
			start := i
			for i < n {
				c = src[i]
				if c == '\n' {
					line++
					i++
					continue
				}
				if c == '\\' && i+1 < n {
					i += 2
					continue
				}
				if c == '`' {
					t.parts = append(t.parts, src[start:i])
					flushTemplate(t)
					stack = stack[:len(stack)-1]
					i++
					break
				}
				if c == '$' && i+1 < n && src[i+1] == '{' {
					t.parts = append(t.parts, src[start:i])
					t.braces = append(t.braces, 0)
					i += 2
					break
				}
				i++
			}
			if i >= n && (len(stack) > 0 && len(stack[len(stack)-1].braces) == 0) {
				// EOF inside a template raw part: close it.
				t := stack[len(stack)-1]
				t.parts = append(t.parts, src[start:])
				flushTemplate(t)
				stack = stack[:len(stack)-1]
			}
			continue
		}

		// code mode (top level or inside a ${ } span).
		switch {
		case c == '\n':
			line++
			i++
		case c == '/' && i+1 < n && src[i+1] == '/':
			for i < n && src[i] != '\n' {
				i++
			}
		case c == '/' && i+1 < n && src[i+1] == '*':
			i += 2
			for i+1 < n && !(src[i] == '*' && src[i+1] == '/') {
				if src[i] == '\n' {
					line++
				}
				i++
			}
			i += 2
			if i > n {
				i = n
			}
		case c == '\'' || c == '"':
			raw, startLine := readQuoted(c)
			out = append(out, Literal{Text: stripPlaceholders(raw), Line: startLine})
		case c == '`':
			stack = append(stack, &tmpl{startLine: line})
			i++
		case c == '{':
			if len(stack) > 0 {
				t := stack[len(stack)-1]
				t.braces[len(t.braces)-1]++
			}
			i++
		case c == '}':
			if len(stack) > 0 {
				t := stack[len(stack)-1]
				if t.braces[len(t.braces)-1] == 0 {
					// the ${ } span closes: back to the template's raw part.
					t.braces = t.braces[:len(t.braces)-1]
				} else {
					t.braces[len(t.braces)-1]--
				}
			}
			i++
		default:
			i++
		}
	}

	// EOF with templates still open (unterminated): close them all.
	for len(stack) > 0 {
		t := stack[len(stack)-1]
		flushTemplate(t)
		stack = stack[:len(stack)-1]
	}
	return out
}

// placeholderRe matches a textual ${VAR} placeholder (compose/env-file
// convention) inside a plain string literal.
var placeholderRe = regexp.MustCompile(`\$\{[^}]*\}`)

// stripPlaceholders replaces textual ${VAR} placeholders with the sentinel —
// an env reference is allowed and poisons host adjacency.
func stripPlaceholders(s string) string {
	return placeholderRe.ReplaceAllString(s, sentinel)
}

// ---------------------------------------------------------------------------
// The classifier — the DECLARED rules (arch-fitness.json), pure and total.
// ---------------------------------------------------------------------------

var (
	localhostRe = regexp.MustCompile(`(?i)(^|[^a-z0-9_-])localhost($|[^a-z0-9_-])`)
	ipv4Re      = regexp.MustCompile(`(^|[^0-9.])([0-9]{1,3})\.([0-9]{1,3})\.([0-9]{1,3})\.([0-9]{1,3})($|[^0-9.])`)
)

// ClassifyLiteral applies the closed rules to one extracted literal (its
// ${...} placeholders already sentinel-poisoned). It returns the FIRST
// matching reason in canonical order and whether the literal is hardcoded.
func ClassifyLiteral(text string) (Reason, bool) {
	t := stripPlaceholders(text)

	// 1. localhost / loopback / wildcard bind.
	if localhostRe.MatchString(t) || hasIPv4(t, func(o [4]int) bool {
		return (o[0] == 127) || (o[0] == 0 && o[1] == 0 && o[2] == 0 && o[3] == 0)
	}) {
		return ReasonLocalhost, true
	}
	// 2. any concrete IPv4.
	if hasIPv4(t, func([4]int) bool { return true }) {
		return ReasonIPLiteral, true
	}
	// 3. scheme://host with a concrete host.
	if urlConcreteHost(t) {
		return ReasonURLConcreteHost, true
	}
	// 4. concrete host:port.
	if hostPortConcrete(t) {
		return ReasonHostPort, true
	}
	return "", false
}

// hasIPv4 reports whether t carries an IPv4 dotted-quad whose octets are all
// <= 255 and satisfy pred.
func hasIPv4(t string, pred func([4]int) bool) bool {
	for _, m := range ipv4Re.FindAllStringSubmatch(t, -1) {
		var o [4]int
		ok := true
		for k := 0; k < 4; k++ {
			v, err := strconv.Atoi(m[2+k])
			if err != nil || v > 255 {
				ok = false
				break
			}
			o[k] = v
		}
		if ok && pred(o) {
			return true
		}
	}
	return false
}

// urlConcreteHost reports whether t contains scheme://host where the host
// part (authority minus userinfo and port) carries a concrete alphanumeric
// character AND no env sentinel.
func urlConcreteHost(t string) bool {
	rest := t
	for {
		idx := strings.Index(rest, "://")
		if idx < 0 {
			return false
		}
		authority := rest[idx+3:]
		if cut := strings.IndexAny(authority, "/?#"); cut >= 0 {
			authority = authority[:cut]
		}
		if at := strings.LastIndex(authority, "@"); at >= 0 {
			authority = authority[at+1:]
		}
		host := authority
		if colon := strings.Index(host, ":"); colon >= 0 {
			host = host[:colon]
		}
		if !strings.Contains(host, sentinel) && strings.ContainsFunc(host, isAlnum) {
			return true
		}
		rest = rest[idx+3:]
	}
}

func isAlnum(r rune) bool {
	return (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9')
}

func isHostChar(b byte) bool {
	return (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || (b >= '0' && b <= '9') ||
		b == '.' || b == '-' || b == sentinel[0]
}

// hostPortConcrete reports whether t contains host:port (port = 2..5 digits)
// where the host token has at least one letter and no env sentinel — the
// concrete service endpoint bypassing the ${APP_NAME}-<svc> convention.
func hostPortConcrete(t string) bool {
	for i := 0; i < len(t); i++ {
		if t[i] != ':' {
			continue
		}
		// digits after the colon.
		j := i + 1
		for j < len(t) && t[j] >= '0' && t[j] <= '9' {
			j++
		}
		digits := j - i - 1
		if digits < 2 || digits > 5 {
			continue
		}
		if j < len(t) && isHostChar(t[j]) {
			continue // the "port" continues into a host-ish token — not a port
		}
		// host token before the colon.
		k := i
		for k > 0 && isHostChar(t[k-1]) {
			k--
		}
		host := t[k:i]
		if host == "" || strings.Contains(host, sentinel) {
			continue
		}
		// scheme-relative authorities are rule 3's business.
		if k >= 2 && t[k-1] == '/' && t[k-2] == '/' {
			continue
		}
		hasLetter := false
		for _, r := range host {
			if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') {
				hasLetter = true
				break
			}
		}
		if hasLetter {
			return true
		}
	}
	return false
}

// ---------------------------------------------------------------------------
// The sensor — Sense / Block / Address (pure, content-addressed).
// ---------------------------------------------------------------------------

// ScanSource runs the literal pass + the classifier over ONE emitted file
// and returns its findings sorted (line, literal).
func ScanSource(path, src string) []Finding {
	var out []Finding
	for _, lit := range ExtractLiterals(src) {
		if reason, hard := ClassifyLiteral(lit.Text); hard {
			out = append(out, Finding{
				File:    path,
				Line:    lit.Line,
				Literal: strings.ReplaceAll(lit.Text, sentinel, "${…}"),
				Reason:  reason,
			})
		}
	}
	sort.Slice(out, func(i, j int) bool {
		if out[i].Line != out[j].Line {
			return out[i].Line < out[j].Line
		}
		return out[i].Literal < out[j].Literal
	})
	return out
}

// Sense is THE DP08 sensor: the deterministic verdict over a whole emitted
// tree (path → source). Green iff NO literal in NO file is a hardcoded
// endpoint. Total and pure — same tree ⇒ same verdict, same address.
func Sense(files map[string]string) (Verdict, error) {
	paths := make([]string, 0, len(files))
	for p := range files {
		paths = append(paths, p)
	}
	sort.Strings(paths)

	findings := []Finding{}
	for _, p := range paths {
		findings = append(findings, ScanSource(p, files[p])...)
	}
	state := StateGreen
	if len(findings) > 0 {
		state = StateRed
	}
	treeAddr, err := canonicalAddress(map[string]any{"files": files})
	if err != nil {
		return Verdict{}, err
	}
	return Verdict{
		Rule:        string(CodeEmittedNoHardcodedEndpoint),
		State:       state,
		TreeAddress: treeAddr,
		Findings:    findings,
	}, nil
}

// Canonical renders the verdict as S02-canonical bytes (records.Canonicalize
// — keys sorted, no insignificant whitespace): the body the TS twin
// reproduces byte-for-byte.
func Canonical(v Verdict) ([]byte, error) {
	body, err := json.Marshal(map[string]any{"verdict": v})
	if err != nil {
		return nil, err
	}
	return records.Canonicalize(body)
}

// Address is the content address of the canonical verdict (records.Hash —
// S02 reused). The Go address is AUTHORITATIVE; the vitest twin and the
// Playwright e2e pin it (« même arbre → même verdict »).
func Address(v Verdict) (string, error) {
	canon, err := Canonical(v)
	if err != nil {
		return "", err
	}
	return records.Hash(canon), nil
}

func canonicalAddress(body map[string]any) (string, error) {
	raw, err := json.Marshal(body)
	if err != nil {
		return "", err
	}
	canon, err := records.Canonicalize(raw)
	if err != nil {
		return "", err
	}
	return records.Hash(canon), nil
}

// Block turns a red verdict into the actionable EMITTED_NO_HARDCODED_ENDPOINT
// BlockReason that BLOCKS THE CUT; a green verdict blocks nothing (nil).
func Block(v Verdict) *blockreason.BlockReason {
	if v.State != StateRed {
		return nil
	}
	first := v.Findings[0]
	return &blockreason.BlockReason{
		Code:     CodeEmittedNoHardcodedEndpoint,
		Severity: blockreason.SeverityBlocking,
		Explanation: fmt.Sprintf(
			"Le source émis porte %d endpoint(s) en dur — le premier : %s:%d %q (%s). "+
				"SPEC-stack-2026 : aucune URL, host, IP ou port en dur dans le source émis ; "+
				"tout endpoint passe par la projection DP07 (resolveConnection → process.env au boot).",
			len(v.Findings), first.File, first.Line, first.Literal, first.Reason),
		HowToFix: []string{
			"route_endpoint : remplacez le littéral par la projection DP07 — resolveConnection(service, env) → requireEnv/process.env au boot (jamais une valeur dans le source).",
			"re_emit : ré-émettez le module de connexions (EmitConnectionsModule) pour l'environnement visé — il ne porte que des références ${VAR}.",
			"rerun aidos check : rejouez le senseur — la coupe se débloque quand l'arbre émis est vert.",
		},
	}
}

// SensorArchFit projects the DP08 verdict into the S84 archfit sensor slot
// (selfcert — the auto-certification battery): a red endpoint verdict reddens
// archfit and the battery blocks the cut. FAIL-CLOSED: a tree the sensor
// cannot canonicalize is RED, never silently green (anti-passthrough).
func SensorArchFit(files map[string]string) selfcert.SensorVerdict {
	v, err := Sense(files)
	if err != nil {
		return selfcert.SensorVerdict{
			Kind:   selfcert.SensorArchFit,
			State:  selfcert.SensorRed,
			Detail: fmt.Sprintf("%s: unscannable emitted tree (fail-closed): %v", CodeEmittedNoHardcodedEndpoint, err),
		}
	}
	if v.State == StateRed {
		f := v.Findings[0]
		return selfcert.SensorVerdict{
			Kind:  selfcert.SensorArchFit,
			State: selfcert.SensorRed,
			Detail: fmt.Sprintf("%s: %d hardcoded endpoint(s) — first %s:%d %q (%s)",
				CodeEmittedNoHardcodedEndpoint, len(v.Findings), f.File, f.Line, f.Literal, f.Reason),
		}
	}
	return selfcert.SensorVerdict{Kind: selfcert.SensorArchFit, State: selfcert.SensorGreen}
}

// ---------------------------------------------------------------------------
// The demo sandbox tree — the canonical fixture the Workbench /endpoints-
// fitness screen, the vitest twin and the Playwright e2e all replay. BYTE-
// IDENTICAL to DEMO_EMITTED_TREE in front/web/lib/endpoint-fitness.ts (the
// pinned addresses red on one byte of drift).
// ---------------------------------------------------------------------------

// LeakPath is the canonical fault-injection file path.
const LeakPath = "gen/app/leak.ts"

// LeakSource is the canonical injected hardcoded endpoint (the DP08
// done-criterion example: https://1.2.3.4:5432).
const LeakSource = "// FAULT INJECTION (DP08 sandbox) — the canonical hardcoded endpoint.\nexport const LEAK = \"https://1.2.3.4:5432\";\n"

// DemoEmittedTree returns a fresh copy of the canonical GREEN emitted tree:
// an emitted-style DP07 boot-config module + a worker consuming it. Every
// endpoint is env-derived — the sensor must hold it green.
func DemoEmittedTree() map[string]string {
	return map[string]string{
		"gen/app/connections.prod.ts": "// Code generated by AIDOS connresolve (DP07) — DO NOT EDIT.\n" +
			"function requireEnv(name: string): string {\n" +
			"\tconst v = process.env[name];\n" +
			"\tif (v === undefined || v === \"\") {\n" +
			"\t\tthrow new Error(`MISSING_ENV_AT_BOOT: ${name}`);\n" +
			"\t}\n" +
			"\treturn v;\n" +
			"}\n" +
			"\n" +
			"export const connections = {\n" +
			"\tcache: `${requireEnv(\"APP_NAME\")}-cache:6379`,\n" +
			"\tcrm: requireEnv(\"CRM_MANAGED_URL\"),\n" +
			"\tdb: `${requireEnv(\"APP_NAME\")}-db:5432`,\n" +
			"\tserver: `https://${requireEnv(\"APP_SUBDOMAIN\")}.${requireEnv(\"DOMAIN\")}`,\n" +
			"};\n",
		"gen/app/worker.ts": "// Code generated by AIDOS (DP08 demo worker) — DO NOT EDIT.\n" +
			"import { connections } from \"./connections.prod\";\n" +
			"\n" +
			"export function queueTarget(): string {\n" +
			"\treturn connections.cache;\n" +
			"}\n",
	}
}

File: front/web/lib/endpoint-fitness.ts
import { createHash } from "node:crypto";
import ts from "typescript";

/**
 * endpoint-fitness — the DP08 arch-fitness invariant
 * EMITTED_NO_HARDCODED_ENDPOINT as the TS AST PASS (A7, ADR 0066): the REAL
 * TypeScript compiler API walks the emitted tree (Hono handlers, workers,
 * boot config) and classifies every string/template literal with the
 * DECLARED closed rules — no host, URL, IP or concrete host:port may live
 * as a literal in emitted source; every endpoint flows through the DP07
 * projection (resolveConnection → process.env / requireEnv at boot —
 * SPEC-stack-2026: « AUCUNE URL/secret en dur »).
 *
 * DETERMINISM-FIRST: a pure function of the tree — same tree → same verdict
 * → same content address (the reproducibility mirror). NEVER an LLM judge.
 * The authoritative Go twin is back/runtime/endpointfitness (its lexer has
 * AST-equivalent semantics); the pinned addresses in endpoint-fitness.test.ts
 * red on one byte of divergence. The declared rule lives above the line in
 * back/runtime/agentloop/arch-fitness.json (changing it is idée → miroir →
 * /goal). THE WALL: this module measures — it writes nothing.
 */

/** RULE — the arch-fitness invariant name and BlockReason code. */
export const RULE = "EMITTED_NO_HARDCODED_ENDPOINT";

/** The CLOSED reason set, canonical order (parity with Go Reasons()). */
export const REASONS = [
	"localhost_literal",
	"ip_literal",
	"url_concrete_host",
	"host_port_literal",
] as const;
export type Reason = (typeof REASONS)[number];

export interface Finding {
	file: string;
	line: number;
	literal: string;
	reason: Reason;
}

export interface Verdict {
	rule: string;
	state: "green" | "red";
	tree_address: string;
	findings: Finding[];
}

// sentinel poisons adjacency where a template ${expr} span or a textual
// ${VAR} placeholder stood — an env-derived fragment never completes a
// concrete host. NUL never appears in real TS source.
const SENTINEL = "\u0000";

const PLACEHOLDER_RE = /\$\{[^}]*\}/g;

function stripPlaceholders(s: string): string {
	return s.replace(PLACEHOLDER_RE, SENTINEL);
}

// ---------------------------------------------------------------------------
// The classifier — the DECLARED rules (arch-fitness.json), pure and total.
// ---------------------------------------------------------------------------

const LOCALHOST_RE = /(^|[^a-z0-9_-])localhost($|[^a-z0-9_-])/i;
const IPV4_RE =
	/(^|[^0-9.])([0-9]{1,3})\.([0-9]{1,3})\.([0-9]{1,3})\.([0-9]{1,3})($|[^0-9.])/g;

function hasIPv4(t: string, pred: (o: number[]) => boolean): boolean {
	for (const m of t.matchAll(IPV4_RE)) {
		const octets = [m[2], m[3], m[4], m[5]].map((s) => Number(s));
		if (octets.every((v) => v <= 255) && pred(octets)) return true;
	}
	return false;
}

function isAlnum(c: string): boolean {
	return /[a-zA-Z0-9]/.test(c);
}

function urlConcreteHost(t: string): boolean {
	let rest = t;
	for (;;) {
		const idx = rest.indexOf("://");
		if (idx < 0) return false;
		let authority = rest.slice(idx + 3);
		const cut = authority.search(/[/?#]/);
		if (cut >= 0) authority = authority.slice(0, cut);
		const at = authority.lastIndexOf("@");
		if (at >= 0) authority = authority.slice(at + 1);
		let host = authority;
		const colon = host.indexOf(":");
		if (colon >= 0) host = host.slice(0, colon);
		if (!host.includes(SENTINEL) && [...host].some(isAlnum)) return true;
		rest = rest.slice(idx + 3);
	}
}

function isHostChar(c: string): boolean {
	return /[a-zA-Z0-9.-]/.test(c) || c === SENTINEL;
}

function hostPortConcrete(t: string): boolean {
	for (let i = 0; i < t.length; i++) {
		if (t.charAt(i) !== ":") continue;
		let j = i + 1;
		while (j < t.length && t.charAt(j) >= "0" && t.charAt(j) <= "9") j++;
		const digits = j - i - 1;
		if (digits < 2 || digits > 5) continue;
		if (j < t.length && isHostChar(t.charAt(j))) continue;
		let k = i;
		while (k > 0 && isHostChar(t.charAt(k - 1))) k--;
		const host = t.slice(k, i);
		if (host === "" || host.includes(SENTINEL)) continue;
		// scheme-relative authorities are the URL rule's business.
		if (k >= 2 && t[k - 1] === "/" && t[k - 2] === "/") continue;
		if (/[a-zA-Z]/.test(host)) return true;
	}
	return false;
}

/**
 * classifyLiteral — the closed rules over one extracted literal. Returns the
 * FIRST matching reason in canonical order, or null when the literal passes.
 */
export function classifyLiteral(text: string): Reason | null {
	const t = stripPlaceholders(text);
	if (
		LOCALHOST_RE.test(t) ||
		hasIPv4(t, (o) => o[0] === 127 || o.every((v) => v === 0))
	) {
		return "localhost_literal";
	}
	if (hasIPv4(t, () => true)) return "ip_literal";
	if (urlConcreteHost(t)) return "url_concrete_host";
	if (hostPortConcrete(t)) return "host_port_literal";
	return null;
}

// ---------------------------------------------------------------------------
// The AST pass — the real TypeScript compiler API (A7).
// ---------------------------------------------------------------------------

interface ExtractedLiteral {
	text: string;
	line: number;
}

function lineOf(sf: ts.SourceFile, node: ts.Node): number {
	return sf.getLineAndCharacterOfPosition(node.getStart(sf)).line + 1;
}

function rawOf(sf: ts.SourceFile, node: ts.TemplateLiteralLikeNode): string {
	if (node.rawText !== undefined) return node.rawText;
	const text = node.getText(sf);
	return text.slice(1, -1);
}

/**
 * extractLiterals — every string literal (its textual ${VAR} placeholders
 * sentinel-poisoned), every template literal (its raw parts joined with the
 * sentinel — the TemplateExpression semantics) and, exactly like the Go
 * twin's lexer, the literals found INSIDE template ${expr} spans.
 */
export function extractLiterals(src: string): ExtractedLiteral[] {
	const sf = ts.createSourceFile(
		"emitted.ts",
		src,
		ts.ScriptTarget.Latest,
		true,
		ts.ScriptKind.TS,
	);
	const out: ExtractedLiteral[] = [];

	const visit = (node: ts.Node): void => {
		if (ts.isStringLiteral(node)) {
			const raw = node.getText(sf).slice(1, -1);
			out.push({ text: stripPlaceholders(raw), line: lineOf(sf, node) });
			return;
		}
		if (ts.isNoSubstitutionTemplateLiteral(node)) {
			out.push({ text: rawOf(sf, node), line: lineOf(sf, node) });
			return;
		}
		if (ts.isTemplateExpression(node)) {
			const parts = [rawOf(sf, node.head)];
			for (const span of node.templateSpans)
				parts.push(rawOf(sf, span.literal));
			out.push({ text: parts.join(SENTINEL), line: lineOf(sf, node) });
			// the literals inside ${expr} spans are still scanned (AST walk).
			for (const span of node.templateSpans) visit(span.expression);
			return;
		}
		ts.forEachChild(node, visit);
	};
	visit(sf);
	return out;
}

/** scanSource — the pass + the classifier over ONE emitted file, sorted. */
export function scanSource(path: string, src: string): Finding[] {
	const out: Finding[] = [];
	for (const lit of extractLiterals(src)) {
		const reason = classifyLiteral(lit.text);
		if (reason !== null) {
			out.push({
				file: path,
				line: lit.line,
				// biome-ignore lint/suspicious/noTemplateCurlyInString: the sentinel is DISPLAYED as a literal ${…} env-reference marker (Go twin parity).
				literal: lit.text.replaceAll(SENTINEL, "${…}"),
				reason,
			});
		}
	}
	out.sort((a, b) =>
		a.line !== b.line
			? a.line - b.line
			: a.literal < b.literal
				? -1
				: a.literal > b.literal
					? 1
					: 0,
	);
	return out;
}

// ---------------------------------------------------------------------------
// Canonicalization + content address (records.Canonicalize/Hash twin — S02).
// ---------------------------------------------------------------------------

function encodeCanonical(v: unknown): string {
	if (v === null) return "null";
	if (Array.isArray(v)) {
		return `[${v.map(encodeCanonical).join(",")}]`;
	}
	if (typeof v === "object") {
		const keys = Object.keys(v as Record<string, unknown>).sort();
		const parts = keys.map(
			(k) =>
				`${JSON.stringify(k)}:${encodeCanonical((v as Record<string, unknown>)[k])}`,
		);
		return `{${parts.join(",")}}`;
	}
	return JSON.stringify(v);
}

function sha256hex(s: string): string {
	return createHash("sha256").update(s, "utf8").digest("hex");
}

/**
 * sense — THE DP08 sensor: the deterministic verdict over a whole emitted
 * tree (path → source). Green iff NO literal in NO file is a hardcoded
 * endpoint. Pure: same tree → same verdict.
 */
export function sense(files: Record<string, string>): Verdict {
	const paths = Object.keys(files).sort();
	const findings: Finding[] = [];
	for (const p of paths) {
		findings.push(...scanSource(p, files[p] ?? ""));
	}
	const treeAddress = sha256hex(encodeCanonical({ files }));
	return {
		rule: RULE,
		state: findings.length > 0 ? "red" : "green",
		tree_address: treeAddress,
		findings,
	};
}

/** canonicalVerdict — the S02-canonical bytes of a verdict (Go twin parity). */
export function canonicalVerdict(v: Verdict): string {
	return encodeCanonical({ verdict: v });
}

/**
 * hashVerdict — the content address of the canonical verdict. MUST equal the
 * Go-authoritative address on the demo tree (pinned in the vitest mirror and
 * the Playwright e2e) — « même arbre → même verdict ».
 */
export function hashVerdict(v: Verdict): string {
	return sha256hex(canonicalVerdict(v));
}

// ---------------------------------------------------------------------------
// The demo sandbox tree — BYTE-IDENTICAL to Go DemoEmittedTree()
// (back/runtime/endpointfitness): the canonical fixture the /endpoints-
// fitness screen, the vitest mirror and the Playwright e2e all replay.
// ---------------------------------------------------------------------------

/** LEAK_PATH — the canonical fault-injection file path. */
export const LEAK_PATH = "gen/app/leak.ts";

/** LEAK_SOURCE — the canonical injected hardcoded endpoint (DP08 example). */
export const LEAK_SOURCE =
	'// FAULT INJECTION (DP08 sandbox) — the canonical hardcoded endpoint.\nexport const LEAK = "https://1.2.3.4:5432";\n';

/** DEMO_EMITTED_TREE — the canonical GREEN emitted tree (env-derived only). */
export const DEMO_EMITTED_TREE: Record<string, string> = {
	"gen/app/connections.prod.ts":
		"// Code generated by AIDOS connresolve (DP07) — DO NOT EDIT.\n" +
		"function requireEnv(name: string): string {\n" +
		"\tconst v = process.env[name];\n" +
		'\tif (v === undefined || v === "") {\n' +
		// biome-ignore lint/suspicious/noTemplateCurlyInString: the fixture IS emitted TS source — byte-identical to Go DemoEmittedTree().
		"\t\tthrow new Error(`MISSING_ENV_AT_BOOT: ${name}`);\n" +
		"\t}\n" +
		"\treturn v;\n" +
		"}\n" +
		"\n" +
		"export const connections = {\n" +
		// biome-ignore lint/suspicious/noTemplateCurlyInString: emitted fixture source — env-derived endpoint, byte-identical to Go.
		'\tcache: `${requireEnv("APP_NAME")}-cache:6379`,\n' +
		'\tcrm: requireEnv("CRM_MANAGED_URL"),\n' +
		// biome-ignore lint/suspicious/noTemplateCurlyInString: emitted fixture source — env-derived endpoint, byte-identical to Go.
		'\tdb: `${requireEnv("APP_NAME")}-db:5432`,\n' +
		// biome-ignore lint/suspicious/noTemplateCurlyInString: emitted fixture source — env-derived endpoint, byte-identical to Go.
		'\tserver: `https://${requireEnv("APP_SUBDOMAIN")}.${requireEnv("DOMAIN")}`,\n' +
		"};\n",
	"gen/app/worker.ts":
		"// Code generated by AIDOS (DP08 demo worker) — DO NOT EDIT.\n" +
		'import { connections } from "./connections.prod";\n' +
		"\n" +
		"export function queueTarget(): string {\n" +
		"\treturn connections.cache;\n" +
		"}\n",
};
